library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java

```diff
@@ -53,41 +50,87 @@
   long size();
 
   /**
-   * Sets the data associated with the given window
-   *
-   * @param window
-   * @param value
-   */
-  void put(Window window, T value);
-
-  /**
-   * Gets the value associated with the given window
-   *
-   * @param window
-   * @return
-   */
-  T get(Window window);
-
-  /**
    * Removes all the data associated with the given window. This does NOT mean removing the window in checkpointed state
    *
    * @param window
    */
   void remove(Window window);
 
   /**
-   * Migrate the data from one window to another. This will invalidate fromWindow in the storage and move the
-   * data to toWindow, and overwrite any existing data in toWindow
+   * This interface handles plain value per window. If there is a key/value map for each window, use
+   * {@link WindowedKeyedStorage}. Also note that a single T object is assumed to be fit in memory
+   *
+   * Note that this interface expects that the implementation takes care of checkpoint recovery.
    *
-   * @param fromWindow
-   * @param toWindow
+   * @param <T> The type of the data that is stored per window
    */
-  void migrateWindow(Window fromWindow, Window toWindow);
+  interface WindowedPlainStorage<T> extends WindowedStorage, Iterable<Map.Entry<Window, T>>
+  {
+    /**
+     * Sets the data associated with the given window
+     *
+     * @param window
+     * @param value
+     */
+    void put(Window window, T value);
+
+    /**
+     * Gets the value associated with the given window
+     *
+     * @param window
+     * @return
+     */
+    T get(Window window);
+
+    /**
+     * Returns the iterable of the entries in the storage
+     *
+     * @return
+     */
+    Iterable<Map.Entry<Window, T>> entries();
```

Comment:
Actually I'd rather change the method to allWindows and remove Iterable from class signature. It's hard to understand a WindowedPlainStorage is iterable.